src/librustc_data_structures/sharded.rs

```diff
@@ -29,21 +30,36 @@ pub struct Sharded<T> {
 impl<T: Default> Default for Sharded<T> {
     #[inline]
     fn default() -> Self {
+        Self::new(|| T::default())
+    }
+}
+
+impl<T> Sharded<T> {
+    #[inline]
+    pub fn new(mut value: impl FnMut() -> T) -> Self {
+        // Create a vector of the values we want
+        let mut values: SmallVec<[_; SHARDS]> = (0..SHARDS).map(|_| {
```

Comment:
This is fine I guess, but more complex than I'd like. A lot of work to avoid a `Vec`! How often do we use this `Sharded` type... seems like it's only instantiated a few times. But I guess the point is more to avoid the indirection on each use. Maybe worth benchmarking.